### PR TITLE
Separate distance calculation

### PR DESCRIPTION
**Pull request by akokam**
Waits for #79 to be merged first!

### CODE REVIEW

File: co2calculator/calculate.py [hunk 1]
```diff
@@ -764,3 +663,105 @@ def commuting_emissions_group(
...
+
+        coords = []
+        for loc in [start, destination]:
+            _, _, loc_coords, _ = geocoding_structured(loc)
+            coords.append(loc_coords)
+        return get_route(coords, "driving-car")
+
+    if transportation_mode == "bus":
+        # TODO: Validate with BaseModel
+        # TODO: Question: Why are we not calculating the bus trip like `driving-car` routes?
```
Comment (han16nah) [inline co2calculator/calculate.py:708]: I think we are assuming line buses here, which do not go the fastest way between places and that's why we rather take the haversine + detour approach.. with a detour value we found in the literature

Although I'm not completely convinced. For travel buses which are not line buses, get_route would be perfectly fine... Or also get_route + a detour could be an option; OR: Take the example of flixbus: Users could also add the intermediate stops between their start and destination to approximate the route (with get_route)

So I'd say let's discuss this in a next meeting ;)

File: co2calculator/calculate.py [hunk 2]
```diff
@@ -338,38 +260,31 @@ def calc_co2_train(
...
 
 def calc_co2_plane(
-    start: str, destination: str, seating_class: str = None
+    distance: Kilometer, seating_class: str = None
 ) -> Tuple[Kilogram, Kilometer]:
     """
     Function to compute emissions of a plane trip
-    :param start: IATA code of start airport
-    :param destination: IATA code of destination airport
+    :param distance: Distance if plane flight
```
Comment (han16nah) [inline co2calculator/calculate.py:267]: of

File: co2calculator/calculate.py [hunk 3]
```diff
@@ -338,38 +260,31 @@ def calc_co2_train(
...
     :param seating_class: Seating class in the airplane; Emission factors differ between seating classes because
                           business class or first class seats take up more space. An airplane with more such therefore
                           needs to have higher capacity to transport less people -> more co2
                           ["average", "economy_class", "business_class", "premium_economy_class", "first_class"]
-    :type start: str
-    :type destination: str
+    :type destination: Kilometer
```
Comment (han16nah) [inline co2calculator/calculate.py:272]: `:type distance: Kilometer` ?

File: co2calculator/calculate.py [hunk 4]
```diff
@@ -410,17 +327,13 @@ def calc_co2_plane(
...
     Function to compute emissions of a ferry trip
-    :param start: dictionary of location of start port,
-                        e.g., in the form {"locality":<city>, "county":<country>}
-    :param destination: dictionary of location of destination port,
-                        e.g., in the form {"locality":<city>, "county":<country>}
+    :param destination: Distance of ferry trip
     :param seating_class: ["average", "Foot passenger", "Car passenger"]
-    :type start: dict
-    :type destination: dict
+    :type destination: Kilometer
```
Comment (han16nah) [inline co2calculator/calculate.py:336]: `:type distance: Kilometer`

File: co2calculator/calculate.py [hunk 5]
```diff
@@ -585,62 +493,50 @@ def calc_co2_businesstrip(
...
+    # Rules:
+    # - `distance` is dominant;
+    # - if distance not provided, take stops;
+    # - if stops not available, raise error;
+    #
+    # In general:
+    # - If stop-based, calculate distance first, then continue only distance-based
+
+    if not distance:
+        distance = _get_distance(start, destination, transportation_mode)
```
Comment (han16nah) [inline co2calculator/calculate.py:508]: awesome 👍

File: co2calculator/calculate.py [hunk 6]
```diff
@@ -305,28 +247,8 @@ def calc_co2_train(
         warnings.warn(
```
Comment (han16nah) [inline co2calculator/calculate.py:247]: Just spotted a small mistake in L 243:
should be "Train" not "Car": 
`f"Train fuel type was not provided. Using default value: '{fuel_type}'"`

File: tests/unit/test_calculate.py [hunk 7]
```diff
@@ -294,121 +195,45 @@ def test_train__distance_based(
...
     ],
 )
 def test_calc_co2_plane(
-    mocker: MockerFixture,
+    distance: float,
     seating_class: Optional[str],
-    mocked_distance: float,
     expected_emissions: float,
 ):
     """Test: Calculate plane-trip emissions based on start and destination.
```
Comment (han16nah) [inline tests/unit/test_calculate.py:219]: based on given distance

File: tests/unit/test_calculate.py [hunk 8]
```diff
@@ -421,34 +246,17 @@ def test_calc_co2_plane__failed(mocker: MockerFixture):
         # pytest.param("Car passenger", 1, id="seating_class: 'Car passenger"),
     ],
 )
-def test_calc_ferry(
-    mocker: MockerFixture, seating_class: Optional[str], expected_emissions: float
-):
+def test_calc_ferry(seating_class: Optional[str], expected_emissions: float) -> None:
     """Test: Calculate ferry-trip emissions based on start and destination.
```
Comment (han16nah) [inline tests/unit/test_calculate.py:250]: based on given distance

**Review by han16nah — COMMENTED** [review]
Looks very good to me! :)
Tests all passed and I really like that we now have a proper separation of unit tests and functional tests and avoid the ORS calls in the unit tests, nice job :)

Just have some minor suggestions (nothing that affects functionality)

File: co2calculator/calculate.py (hunk 1 above)
Comment (akokam) [reply co2calculator/calculate.py]: #82

**Comment by akokam** [comment]
Okay, this branch had the goal to truly separate the distance calculations from emissions. That's done!

However, I've added more than the scope of #70 was. The later-commits-refactoring of `get_distance` was a little overshoot and should have waited for #82. I really hope it's not too big of a deal for reviewing. I apologize!

File: co2calculator/distances.py [hunk 9]
```diff
@@ -308,3 +355,169 @@ def get_route(coords, profile: str = None) -> Kilometer:
...
+        detour_constant = detour_df[
+            detour_df["transportation_mode"] == transportation_mode
+        ]["constant [km]"].values[0]
+    except KeyError:
+        detour_coefficient = 1.0
+        detour_constant = 0.0
+        warnings.warn(
+            f"""
+        No detour coefficient or constant available for this transportation mode.
+        Detour parameters are available for the following transportation modes:
```
Comment (han16nah) [inline co2calculator/distances.py:383]: Here, we might print out detour_df["transportation_mode"] - or we just leave this out in the warning.

**Review by han16nah — APPROVED** [review]
Awesome, great job! :)

**Pull request merged**